lua/starfall/sflib.lua

```diff
@@ -701,48 +675,230 @@ if SERVER then
-			net.Start( "starfall_upload" )
-			net.WriteBit( true )
-			net.WriteString( list.mainfile )
-			net.SendToServer()
-			--print( "\tHeader sent" )
-
-			local fname = next( list.files )
-			while fname do
-				--print( "\tSending data for:", fname )
-				local fdata = list.files[ fname ]
-				local offset = 1
-				repeat
-					net.Start( "starfall_upload" )
-						net.WriteBit( false )
-						net.WriteString( fname )
-						local data = fdata:sub( offset, offset + 60000 )
-						net.WriteString( data )
-					net.SendToServer()
-
-					--print( "\t\tSent data from", offset, "to", offset + #data )
-					offset = offset + #data + 1
-				until offset > #fdata
-				fname = next( list.files, fname )
-			end
+		if not ok then
+			currentlist = nil
+			return
+		else
+			currentlist = list
+		end
 
-			net.Start( "starfall_upload" )
-				net.WriteBit( true )
+		--print("Server requested code upload, sending file index.")
+
+		local count = table.Count(currentlist.files)
+
+		net.Start( "starfall_uploadlist" )
+		net.WriteString(currentlist.mainfile)
+		net.WriteUInt(count, 16)
+		for fname, code in pairs(currentlist.files) do
+			net.WriteString(fname)
+			net.WriteString(currentlist.hashes[fname])
+			net.WriteUInt(#code, 18)
+		end
+		net.SendToServer()
+	end)
+
+	net.Receive( "starfall_requestfiles", function ( len )
+
+		if currentlist == nil then
+			ErrorNoHalt( "Server requested files before building the index." )
+			return
+		end
+
+		--print("Received server request for files.")
+
+		local requested = {}
+
+		local count = net.ReadUInt(16)
+		for i = 1, count do
+			local fname = net.ReadString()
+			local hash = net.ReadString()
+			local codeSize = net.ReadUInt(18)
+			table.insert(requested, { filename = fname, hash = hash, codeSize = codeSize })
+		end
+
+		for _,v in pairs(requested) do
+			-- Chunk info
+			--print("Sending file: " .. v.filename .. "(Hash: " .. v.hash .. ", size: " .. v.codeSize .. ")")
+
+			net.Start("starfall_upload")
+			net.WriteUInt(0, 2) -- Start of stream
+			net.WriteString(v.filename)
+			net.WriteString(v.hash)
+			net.WriteUInt(v.codeSize, 18)
 			net.SendToServer()
-			--print( "Done sending" )
-		else
+
+			-- Send all chunks.
+			local codeData = currentlist.files[v.filename]
+			local offset = 1
+			repeat
+
+				local data = codeData:sub( offset, offset + 50000 )
+
+				--print("Sending chunk (" .. v.filename .. ") offset: " .. tostring(offset) .. ", len: " .. tostring(#data))
+
+				net.Start( "starfall_upload" )
+				net.WriteUInt(1, 2) -- Chunk piece.
+				net.WriteString( v.filename )
+				net.WriteString( data )
+				net.SendToServer()
+
+				offset = offset + #data
+			until offset > #codeData
+
 			net.Start( "starfall_upload" )
-				net.WriteBit( false )
+				net.WriteUInt(2, 2) -- End of file
+				net.WriteString( v.filename )
+				net.WriteString( v.hash )
 			net.SendToServer()
-			if list then
-				SF.AddNotify( LocalPlayer(), list, NOTIFY_ERROR, 7, NOTIFYSOUND_ERROR1 )
```

Comment:
Concerned about this being removed? Do players still get notified?

Reply:
I wasn't exactly sure what this line does it did seem like a mistake of error handling but I guess its not supposed for errors?

Reply:
The documentation for SF.Editor.BuildIncludesTable isn't clear but the function returns an error message as the second return value if it failed to build the includes table.

See here:
https://github.com/INPStarfall/Starfall/blob/develop/lua/starfall/editor.lua#L1502